tests/test_hg.py

```diff
@@ -102,3 +63,92 @@ def test_integrated_hgrepo_can_log(hg_clone):
     repo = HgRepo(hg_clone.strpath)
     with repo:
         assert repo.run_hg_cmds([["log"]])
+
+
+PATCH_WITHOUT_STARTLINE = rb"""
+# HG changeset patch
+# User Test User <test@example.com>
+# Date 0 0
+#      Thu Jan 01 00:00:00 1970 +0000
+add another file.
+diff --git a/test.txt b/test.txt
+--- a/test.txt
++++ b/test.txt
+@@ -1,1 +1,2 @@
+ TEST
++adding another line
+""".strip()
+
+
+PATCH_WITH_CONFLICT = rb"""
+# HG changeset patch
+# User Test User <test@example.com>
+# Date 0 0
+#      Thu Jan 01 00:00:00 1970 +0000
+# Diff Start Line 7
+Add to a file that doesn't exist
+diff --git a/not-real.txt b/not-real.txt
+--- a/not-real.txt
++++ b/not-real.txt
+@@ -1,1 +1,2 @@
+ TEST
++This line doesn't exist
+""".strip()
+
+
+PATCH_ADD_NO_NEWLINE_FILE = rb"""
+# HG changeset patch
+# User Test User <test@example.com>
+# Date 0 0
+#      Thu Jan 01 00:00:00 1970 +0000
+# Diff Start Line 7
+file added
+diff --git a/test-newline-file b/test-newline-file
+new file mode 100644
+--- /dev/null
++++ b/test-newline-file
+@@ -0,0 +1,1 @@
++hello
+\ No newline at end of file
+""".strip()
+
+
+PATCH_NORMAL = rb"""
+# HG changeset patch
+# User Test User <test@example.com>
+# Date 0 0
+#      Thu Jan 01 00:00:00 1970 +0000
+# Diff Start Line 7
+add another file.
+diff --git a/test.txt b/test.txt
+--- a/test.txt
++++ b/test.txt
+@@ -1,1 +1,2 @@
+ TEST
++adding another line
+""".strip()
+
```

Comment:
there's some tests missing from https://github.com/mozilla-conduit/autoland-transplant/blob/master/testing/tests/test-post-autoland-job-from-patch.t

unicode, falling back to patch, and patching a file that doesn't exist

Reply:
Noted.

Reply:
Added unicode test, the other two are actually part of the `PATCH_WITH_CONFLICT` test. Note that we're only testing `apply_patch` for all of theses types of tests. We should probably add more unit tests for the rest of the functionality (e.g. `build_patch_for_revision` and `push`, etc). Or some integration tests that do not involve S3, but I think the latter, at least, can be done as a follow up.

Reply:
>` PATCH_WITH_CONFLICT` only passes through the "falling back to patch" but ultimately still fails and re-raises the exception.

Ah, testing that the patch fallback works is important.  

> Any thoughts on what type of merge conflict could fail hg's default patching but pass with patch?

https://bugzilla.mozilla.org/show_bug.cgi?id=1541181 has an example.

In transplant I took the route of passing in a header which triggers a false `hg import` failure.
This ensures the code path is tested even after hg fixes the particular bug in its internal patch code.
https://github.com/mozilla-conduit/autoland-transplant/blob/master/autoland/transplant.py#L404-L418

Reply:
Thanks, I will look into this!

Reply:
This is fixed now. The test that existed already wasn't actually doing anything -- turns out, in order to reproduce this bug the file has to already be on the repo (and not added via a patch). I updated the test so that the new line + trailing `r` are manually committed, and then the patch to remove the file is applied; also confirmed that the outgoing changes actually contain what is expected.

Reply:
I don't think the test for falling back to patch should rely on any specific hg bug existing.  It would be too easy to upgrade the version of hg and have this test generate a false positive as the test doesn't actually check if the fallback code path was taken.

Reply:
Good point - I added a new test that forces this change, as well as modified the "newline bug" test to be skipped if the mercurial version is changed.